### ISSUE TITLE
Symbolic shape fails symbol_to_source when caching is enabled

### ISSUE THREAD

**Issue by davidberard98** [pytorch/pytorch#127970]
### 🐛 Describe the bug

There's two issues here:
- [ ] First, NT/PT2 stack should "lift" these details to the (@davidberard98)
- [ ] Next, this issue only repros w/ caching - we should make it so that caching path doesn't require this, since the non-caching path also doesn't need it (@masnesral / @oulgen)

Repro:
```python
import torch
import torch._inductor.config

# or as an environment variable, TORCHINDUCTOR_FX_GRAPH_CACHE=1
torch._inductor.config.fx_graph_cache = True

def gen_nt(r):
    values = torch.randn(r, 16)
    offsets = torch.tensor([0, 2, 3, 6, 13, r])
    return torch.nested.nested_tensor_from_jagged(values, offsets)

def fn(nt):
    if nt.values().size(0) % 16 == 0:
        return nt.sin()
    return nt.cos()

torch.compile(fn)(gen_nt(19))
torch.compile(fn)(gen_nt(20))
```

### Versions

pytorch main branch ~jul 3

**Comment by oulgen** [comment]
@bdhirsh says he has a PR that will do the lifting

**Comment by bdhirsh** [comment]
@guilhermeleobas has been working on a PR that should do this, I'm testing it out locally (might require some tweaks): https://github.com/pytorch/pytorch/pull/125941/files

If it does fix it, we can land part of the PR sooner to unblock

**Comment by bdhirsh** [comment]
One thing to add: it seems to me like the reason this broke with the cache but doesn't fail when we turn off the cache is that:

(1) when the cache is enabled, inductor is generating shape guards

(2) when the cache is disabled, dynamo is the one generating shape guards

(3) dynamo's logic for generating shape guards does not rely on the invariant that every symbol is a graph arg

(4) inductor's logic for generating shape guards appears to rely on this invariant.

So independent from making this fix, should we also update inductor to not require this invariant (given that dynamo does not rely on the invariant either?) maybe cc @ezyang

**Comment by ezyang** [comment]
@bdhirsh My initial instinct also was to make inductor rely on tracked fakes (instead of inputs) to determine where to get symbols on, but there is a bit of a layering problem here. Inductor cache would prefer to only have to know about inputs to the inductor graph; in fact, it doesn't care at all about guards on symbols that don't show up in the inputs at all, because it's impossible for those guards to have affected Inductor compilation. Those guards would have affected the FX graph that eventually made its way to Inductor, but we're already keying on that.

So, somewhat incredibly, maybe the correct and easiest fix to do, is to just have Inductor drop all guards that mention symbols that are not in example_inputs! Ordinarily this would be pretty naughty but I think it's 100% sound here.

**Comment by bdhirsh** [comment]
Ah - so in this particular repro example, the guard on `nt.values().size(0) (s0)` is completely unnecessary for inductor to run: since `s0` was effectively only specialized on by dynamo, so only dynamo needs to run the guard.

> maybe the correct and easiest fix to do, is to just have Inductor drop all guards that mention symbols that are not in example_inputs!

However, it's not clear to me that this case lines up directly with "`s0` is not a grapharg in the inductor graph". In particular:

(1) in this repro, dynamo decides to remove `s0` as a grapharg because it sees that it is unused in the graph ([code](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/output_graph.py#L1274))

(2) Just because `s0` is unused in the dynamo graph, there is nothing preventing inductor from guarding on e.g. `s0 < int32_max`, right? My understanding is that if we have some node `intermediate: [s0]` that gets passed into an ATen op, inductor is free to specialize on the size of the input to that ATen op during lowering.

**Comment by bdhirsh** [comment]
I realized that the PR I linked above does not directly fix this problem (it only adds the outer subclass's dims as extra graph inputs inside of AOTAutograd, not the inner tensor's dims).

I can think of a few other ways to solve this problem, interesting in folks' thoughts:

(1) have inductor use SymInts that come from sizes of graph args when generating the guards, instead of **only** looking at inputs. This is easy to do, but has the downside that inductor might save guards that are redundant (that dynamo will already have run anyway).

(2) inductor only actually cares about saving shape guards that were accrued after dynamo ran. So one option would be a way to "snapshot" the guards after dynamo, so we can diff them against the guards we see after inductor runs, and remove the dynamo guards

(3) we could also just "keep `s0` as a graph arg", even though `s0` is not used directly in the graph. This feels bad though (dynamo explicitly prunes `s0` because it is unused

**Comment by bdhirsh** [comment]
Curious what folks think. The patch below "fixes" this bug and is basically (1). I think... (2) seems more satisfying though, since in theory it would be nice to know which shape guards came from dynamo so we know that inductor should not need to run them?
```
diff --git a/torch/_inductor/codecache.py b/torch/_inductor/codecache.py
index c47f0175148..675907c920d 100644
--- a/torch/_inductor/codecache.py
+++ b/torch/_inductor/codecache.py
@@ -729,7 +729,17 @@ class FxGraphCache:
         Get the backed SymInt objects from the input list. Note that we can never
         have guards that depend on unbacked symint.
         """
-        return [s for s in inputs if isinstance(s, torch.SymInt) and has_hint(s)]
+        symints = []
+        def add_int(s):
+            if isinstance(s, torch.SymInt) and has_hint(s):
+                symints.append(s)
+        for x in inputs:
+            if isinstance(x, torch.SymInt):
+                add_int(x)
+            elif isinstance(x, torch.Tensor):
+                for s in x.shape:
+                    add_int(s)
+        return symints
```

**Comment by masnesral** [comment]
@bdhirsh That looks like it could work for the case we were investigating from the outage. But I have another example from torchbench where we guards that mention 's0', but it appears nowhere in the graph inputs:

```
(Pdb) p shape_env.guards
[ShapeGuard(expr=s0 <= 880801, stack=<torch.utils._traceback.CapturedTraceback object at 0x7fdc549a33a0>), ShapeGuard(expr=s0 <= 65535, stack=<torch.utils._traceback.CapturedTraceback object at 0x7fdc64cbc130>), 
<truncated>

(Pdb) p example_inputs
[s1, FakeTensor(..., device='cuda:0', size=(1, 32000, 128), dtype=torch.float16), FakeTensor(..., device='cuda:0', size=(s1, 1, 128), dtype=torch.float16), FakeTensor(..., device='cuda:0', size=(s1, 32001), dtype=torch.float16)]
```

**Comment by masnesral** [comment]
@ezyang would you share that example?

**Comment by ezyang** [comment]
In fact, @laithsakka has already written the example lol https://github.com/pytorch/pytorch/issues/127328

**Comment by masnesral** [comment]
> So, somewhat incredibly, maybe the correct and easiest fix to do, is to just have Inductor drop all guards that mention symbols that are not in example_inputs! Ordinarily this would be pretty naughty but I think it's 100% sound here.

@ezyang I _think_ you're arguing that it's sound for inductor to always drop such guards, independent of the cache?

**Comment by ezyang** [comment]
Yes. Because those guards cannot affect inductor compilation. If they affected inductor compilation, there must be some way for inductor to have made use of the symbols. But we agreed those symbols did not appear anywhere in the inputs (or anywhere else in the graph, for that matter.)

**Comment by bdhirsh** [comment]
looks like dynamo has some logic to not prune unused graph args if they are free symbols - but the way we compute free symbols does not account for free symbols that live on inner tensors of a tensor subclass. Fix here: https://github.com/pytorch/pytorch/pull/128045

**Comment by oulgen** [comment]
https://github.com/pytorch/pytorch/pull/128052 should fix this along with @bdhirsh 's PR

**Comment by masnesral** [comment]
@davidberard98 any idea if this is actually fixed (old issue scrubbing)?
